### ISSUE TITLE
Investigate incremental alternatives to materialized views

### ISSUE THREAD

**Issue by jon-betts** [hypothesis/lms#4532]
Materialized views are good because:

 * They're simple
 * Hard to get mixed up or end up with incomplete data
 * The can be implemented completely separately from the app code

However they have some draw backs:

 * They don't update automatically
 * You have to recreate them from scratch every time
 * This means if they are slow, the data is likely to be out of date

Can we do better with some kind of incremental approach?

## Considerations

 * We want to be able to react quickly to new reporting requirements
 * It's ideal if whatever method we want can be adapted and deployed as quickly as possible
 * The simpler it is and the less it's tangled into our app code the better

##  Potential alternatives

 * We could investigate partial updates, where we only update data from the last time we updated
    * This could run much more frequently
    * There is a potential virtuous circle here, as more frequent updates means there is less to update each time, which could mean they are quicker, which means they can be run more frequently, which means, they run quicker...
    * What's the simplest way we could do this?
 * We could attempt to update in real time?
   * Would this involve getting tangled with app code?
 * Anything else?

**Comment by marcospri** [comment]
One approach could be rollup tables: https://github.com/hypothesis/lms/pull/4525

**Comment by jon-betts** [comment]
I've done a bit of a write up here:

 * https://github.com/hypothesis/numberwang/blob/incremental-updates-explore/examples/incremental_updates/thoughts.md

I've tried this lots of ways, but scanning the annotation table is the killer. At the low end it's about 30 minutes. At the upper it's 50 for a more full fat table.

I tried various tricks for getting the table into more normalized tables, but it tends to result in it taking forever for the first batch.
